hibernate-core/src/main/java/org/hibernate/bytecode/internal/BytecodeProviderInitiator.java

```diff
@@ -32,8 +49,33 @@ public final class BytecodeProviderInitiator implements StandardServiceInitiator
 
 	@Override
 	public BytecodeProvider initiateService(Map<String, Object> configurationValues, ServiceRegistryImplementor registry) {
-		String provider = ConfigurationHelper.getString( BYTECODE_PROVIDER, configurationValues, BYTECODE_PROVIDER_NAME_DEFAULT );
-		return buildBytecodeProvider( provider );
+		final ClassLoaderService classLoaderService = castNonNull( registry.getService( ClassLoaderService.class ) );
+		final Collection<BytecodeProvider> bytecodeProviders = classLoaderService.loadJavaServices( BytecodeProvider.class );
+		return getBytecodeProvider( bytecodeProviders );
+	}
+
+	public static BytecodeProvider getBytecodeProvider(Iterable<BytecodeProvider> bytecodeProviders) {
+		final Iterator<BytecodeProvider> iterator = bytecodeProviders.iterator();
+		if ( !iterator.hasNext() ) {
+			// If no BytecodeProvider service is available, default to the "no-op" enhancer
+			return new org.hibernate.bytecode.internal.none.BytecodeProviderImpl();
+		}
+
+		BytecodeProvider provider = null;
+		while ( iterator.hasNext() ) {
+			final BytecodeProvider next = iterator.next();
+			if ( provider == null ) {
+				provider = next;
+			}
+			else if ( provider instanceof org.hibernate.bytecode.internal.none.BytecodeProviderImpl ) {
```

Comment:
I'm a little confused about the purpose of these checks, could you elaborate please?

Wasn't it simpler to pass the `bytecodeProviders` parameter as a Collection, and check its size?
if:
 0 : use the no-op
 1 : use whatever was found
 >1 : throw an error

Reply:
Isn't the `org.hibernate.bytecode.internal.none.BytecodeProviderImpl` always going to be provided by `hibernate-core` itself? I though if a user wanted to provide a custom implementation that would need to override the default, that's the reason of the checks.

I'm also passing an `Iterable<BytecodeProvider>` parameter to allow passing the result from `ServiceLoader.load( BytecodeProvider.class )` directly to this method.

Reply:
> Isn't the org.hibernate.bytecode.internal.none.BytecodeProviderImpl always going to be provided by hibernate-core itself? I thought if a user wanted to provide a custom implementation that would need to override the default, that's the reason of the checks.

Ah yes I see what you mean now. I was assuming that we wouldn't allow to load a custom implementation for now, as I was expecting we'd first move the ByteBuddy implementation into an optional, separate jar.

But feel free to keep it then, it's nice that people won't have to wait for that additional refactoring step - I'm not expecting many users to want to do that though :)

Reply:
I've already simplified the code to just retrieve the first available implementation, and throw if there are more than one. I don't think anyone would be interested in providing custom implementations anyways, at least in the short term. We can always change this once we decouple the bytecode provider from hibernate-core.